Extend pod resource API response to return the information from memory manager

cynepco3hahue opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

One of the requirements for the memory manager feature promotion to beta is to provide better visibility of the node allocatable memory and memory allocated by a container. For this purpose, we will expose memory manager specified metrics under the pod resource API.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The pod resource API will provide memory manager metrics in the case when the memory manager feature gate enabled and the memory manager policy is static.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

Commit:
Extend pod resource API response to return the memory manager information

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

klueska commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -49,6 +50,14 @@ message ContainerResources {
+// ContainerMemory contains information about devices assigned to a container
+message ContainerMemory {
+    string memory_type = 1;
+    repeated uint64 size = 2;
```
Why is this a repeated field?

cynepco3hahue replied:
Copy-paste mistake.

cynepco3hahue replied:
Fixed.

Commit:
Implement all necessary methods to provide memory manager data under pod resources metrics

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

cynepco3hahue commented:
/kind feature

cynepco3hahue commented:
/priority important-longterm

cynepco3hahue commented:
/triage accepted

cynepco3hahue commented:
/assign @klueska

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

Commit:
e2e node: provide tests for memory manager pod resources metrics

- verify memory manager data returned by `GetAllocatableResources`
- verify pod container memory manager data

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

cynepco3hahue commented:
@fromanirh Yep I thought about it, but in the case of pod resources we do not need integration tests with other managers, and on another side, if I will introduce tests under the pod resource, I will need to bring some configuration related to the memory manager and it will make the test more complicated. So I decided to add it under the memory manager test suite, please feel free to advocate to move them under the pod resources tests.

cynepco3hahue commented:
> First things first: having the e2e tests here is not _bad_ per se. What matters if 1. to have test coverage and 2. to have maintenable tests, and at glance (need to have a deeper review) both these requirements are already addressed. That said, my take is that the podresources part of the memory manager support better fits into `podresources_test.go` because
> 
> 1. having all in `memory_manager_test.go` creates the "siloed feature" feeling. Could very much be just me, but the dark side of having everything self contained is that the feature also can feel a bit disconnected from the rest of the code. The podresources test are intentionally meant to span across various features because the podresources API consumes them all, so it wants to be orthogonal to them.
> 2. Ideally, tests should be as much isolated as possible from each other. Obviously if memory manager (say) misbehave it is highly likely that the podresources API misbehaves as well, but still, the more independent the tests are, the better. This is the reason why the podresources test intentionally re-implement (to some extent, see below) the setup path already seen in cpumanager and topologymanager test: the intent is to not have a hidden dependency between the shared setup/teardown path. Having such a dependency will make unnecessarily narrow the exercised code path. The reimplementation cost is expected to be payed back to the less coupling among tests
> 
> Now: is true that for various reasons e2e tests share common code, and that's debatable IMO. But if the two aforementioned goals (test across feature and try to avoid silos; make tests independent as much as possible between each other) are worthy, and I think they are, we should go in that direction over time.

I totally agree that adding these tests under the pod resources tests is more appropriate, but currently, it will over complicate the logic, because we do not have a separation between setup, test, and teardown phases. So to add some new tests that will configure hugepages, create VMs with relevant memory and hugepages resources will rise the cyclomatic complexity(check if hugepages configured, if yes choose one, if no choose the other, similar to SR-IOV device configuration).
I have plans to refactor a little bit pod resource tests, but for now, I prefer to leave it under the memory manager test suite and move it under the pod resources test suite after refactoring(code freeze is coming).

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-memory-manager

fromanirh commented:
> > First things first: having the e2e tests here is not _bad_ per se. What matters if 1. to have test coverage and 2. to have maintenable tests, and at glance (need to have a deeper review) both these requirements are already addressed. That said, my take is that the podresources part of the memory manager support better fits into `podresources_test.go` because
> > 
> > 1. having all in `memory_manager_test.go` creates the "siloed feature" feeling. Could very much be just me, but the dark side of having everything self contained is that the feature also can feel a bit disconnected from the rest of the code. The podresources test are intentionally meant to span across various features because the podresources API consumes them all, so it wants to be orthogonal to them.
> > 2. Ideally, tests should be as much isolated as possible from each other. Obviously if memory manager (say) misbehave it is highly likely that the podresources API misbehaves as well, but still, the more independent the tests are, the better. This is the reason why the podresources test intentionally re-implement (to some extent, see below) the setup path already seen in cpumanager and topologymanager test: the intent is to not have a hidden dependency between the shared setup/teardown path. Having such a dependency will make unnecessarily narrow the exercised code path. The reimplementation cost is expected to be payed back to the less coupling among tests
> > 
> > Now: is true that for various reasons e2e tests share common code, and that's debatable IMO. But if the two aforementioned goals (test across feature and try to avoid silos; make tests independent as much as possible between each other) are worthy, and I think they are, we should go in that direction over time.
> 
> I totally agree that adding these tests under the pod resources tests is more appropriate, but currently, it will over complicate the logic, because we do not have a separation between setup, test, and teardown phases. So to add some new tests that will configure hugepages, create VMs with relevant memory and hugepages resources will rise the cyclomatic complexity(check if hugepages configured, if yes choose one, if no choose the other, similar to SR-IOV device configuration).
> I have plans to refactor a little bit pod resource tests, but for now, I prefer to leave it under the memory manager test suite and move it under the pod resources test suite after refactoring(code freeze is coming).

Works for me, we can improve later. What matters now is to have good coverage.
Could you just add a TODO in your code perhaps? (and/or file an issue and refer to it?)

cynepco3hahue commented:
> Works for me, we can improve later. What matters now is to have good coverage.
> Could you just add a TODO in your code perhaps? (and/or file an issue and refer to it?)

Done, https://github.com/kubernetes/kubernetes/issues/101945.

swatisehgal commented:
/cc

cynepco3hahue commented:
/priority important-soon

cynepco3hahue commented:
/remove-priority important-longterm

cynepco3hahue commented:
Advance priority because the graduation was approved.

fromanirh commented:
/cc

fromanirh commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -141,10 +160,12 @@ func TestListPodResourcesV1(t *testing.T) {
 			m.On("GetAllocatableCPUs").Return(cpuset.CPUSet{})
 			m.On("GetAllocatableDevices").Return(devicemanager.NewResourceDeviceInstances())
-			server := NewV1PodResourcesServer(m, m, m)
+			m.On("GetAllocatableMemory").Return([]state.Block{})
```
why empty? Don't we need testcase(s) on which we return valid allocatable memory?

cynepco3hahue replied:
It the List tests, so for `GetAllocatableMemory` we just set an empty state. I do not sure if it necessary at all to set it here, but I just followed the device and CPU manager examples.
```
m.On("GetAllocatableCPUs").Return(cpuset.CPUSet{})
m.On("GetAllocatableDevices").Return(devicemanager.NewResourceDeviceInstances())
```

fromanirh replied:
I guess we can start like this and improve coverage later - applies indeed to the other resource types.

fromanirh commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -767,3 +767,24 @@ func findBestHint(hints []topologymanager.TopologyHint) *topologymanager.Topolog
+// GetAllocatableMemory returns the amount of allocatable memory for each NUMA node
+func (p *staticPolicy) GetAllocatableMemory(s state.State) []state.Block {
+	var allocatableMemory []state.Block
+	machineState := s.GetMachineState()
```
this is safe to be called concurrently, right?

cynepco3hahue replied:
Yes, it has a lock inside of it
https://github.com/kubernetes/kubernetes/blob/0abb908638dcea611253ed1dfc6d66e8b31d42d4/pkg/kubelet/cm/memorymanager/state/state_checkpoint.go#L103

ffromani reviewed (commented):
looks straightforward to me, minor questions inside, but looks almost good to me :)

cynepco3hahue commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -49,6 +50,14 @@ message ContainerResources {
+    repeated ContainerMemory memory = 4;
+}
+
+// ContainerMemory contains information about devices assigned to a container
```
Thanks! Another copy-paste mistake.

klueska replied:
Looks like this still needs to be updated?

cynepco3hahue replied:
done

ffromani reviewed (commented):
/lgtm
pretty straightforward extension for memory of what we have already in the podresources API
I think you will need another ack and surely the final approval from @klueska

klueska commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -407,3 +418,13 @@ func getSystemReservedMemory(machineInfo *cadvisorapi.MachineInfo, nodeAllocatab
+
+// GetAllocatableMemory returns the amount of allocatable memory for each NUMA node
+func (m *manager) GetAllocatableMemory() []state.Block {
+	return m.allocatableMemory
```
Is it OK for this to be cached like this? I guess the assumption is that _if_ the amount of allocatable memory was ever changed, then the kubelet would be restarted and this value recalculated.

cynepco3hahue replied:
Exactly, to change the allocatable you should update the `reservedMemory` parameter and it will require the kubelet restart.

klueska reviewed (changes requested):
A few small comments. Otherwise looks good.

cynepco3hahue commented on pkg/kubelet/cm/container_manager.go:
```diff
@@ -224,3 +226,25 @@ func containerDevicesFromResourceDeviceInstances(devs devicemanager.ResourceDevi
 	return respDevs
 }
+
+func containerMemoryFromBlock(blocks []memorymanagerstate.Block) []*podresourcesapi.ContainerMemory {
```
I did not really think about the windows and just followed the `containerDevicesFromResourceDeviceInstances` example, I am very skeptical regarding adding anything related to resource management under the windows and will move this method to `container_manager_linux.go`

cynepco3hahue commented:
/retest

klueska commented:
/lgtm
/approve
